### PR TITLE
Add TPE tests.

### PR DESCRIPTION
**Pull request by ytsmiling**
This PR adds missing tests for TPE samplers and also improves the test-coverage of TPE's Parzen-estimator.

### CODE REVIEW

File: tests/samplers_tests/tpe_tests/test_sampler.py [hunk 1]
```diff
@@ -1,4 +1,13 @@
 import pytest
+import random
+from typing import Callable
+from typing import Dict
+from typing import Optional
+from typing import Union
+from unittest.mock import Mock
+from unittest.mock import patch
```
Comment (crcrpar) [inline tests/samplers_tests/tpe_tests/test_sampler.py:8]: pytest is not standard library.

Comment (ytsmiling) [reply tests/samplers_tests/tpe_tests/test_sampler.py:8]: Thanks for checking the PR! Yes, pytest is not, but [unittest](https://docs.python.org/3/library/unittest.html) is. Related PR: #1121.

Comment (crcrpar) [reply tests/samplers_tests/tpe_tests/test_sampler.py:8]: Nah, this test imports pytest before standard libraries.
https://github.com/optuna/optuna/pull/1126/files#diff-9b7cf9235ed9634dee5bbc85e55f8bb6R1

Comment (ytsmiling) [reply tests/samplers_tests/tpe_tests/test_sampler.py:8]: Oh, wow, I didn't notice the first line. Thank you for catching it.

**Comment by ytsmiling** [comment]
> Could you add periods to comments in `tests/samplers_tests/tpe_tests/test_sampler.py`?
> 
> ref: https://github.com/optuna/optuna/wiki/Coding-Style-Conventions#block-and-inline-comments-are-supposed-to-be-sentences

The latest commit resolved this comment.

File: tests/samplers_tests/tpe_tests/test_sampler.py [hunk 2]
```diff
@@ -20,6 +29,218 @@ def test_hyperopt_parameters(use_hyperband):
...
+    assert sampler.infer_relative_search_space(study, frozen_trial) == {}
+
+
+def test_sample_independent() -> None:
+    study = optuna.create_study()
+    dist = optuna.distributions.UniformDistribution(1.0, 100.0)
+    past_trials = [frozen_trial_factory(i, dist=dist) for i in range(1, 8)]
+
+    # Prepare a trial and a sample for later checks.
+    with patch("optuna.Study.get_trials", return_value=past_trials):
```
Comment (hvy) [inline tests/samplers_tests/tpe_tests/test_sampler.py:54]: Could you place operations that are not dependent on `optuna.Study.get_trials` outside this context manager? Above all, it aids mental readability.

File: tests/samplers_tests/tpe_tests/test_sampler.py [hunk 3]
```diff
@@ -20,6 +29,281 @@ def test_hyperopt_parameters(use_hyperband):
...
+    dist = optuna.distributions.UniformDistribution(1.0, 100.0)
+
+    # Prepare sampling result for later tests.
+    past_trials = [frozen_trial_factory(i, dist=dist) for i in range(1, 30)]
+    trial = frozen_trial_factory(30)
+    sampler = TPESampler(n_startup_trials=5, seed=0)
+    with patch("optuna.Study.get_trials", return_value=past_trials):
+        all_success_suggestion = sampler.sample_independent(study, trial, "param-a", dist)
+
+    # Test unsuccessful trials are handled differently
```
Comment (sile) [inline tests/samplers_tests/tpe_tests/test_sampler.py:255]:
```suggestion
    # Test unsuccessful trials are handled differently.
```

File: tests/samplers_tests/tpe_tests/test_sampler.py [hunk 4]
```diff
@@ -20,6 +29,281 @@ def test_hyperopt_parameters(use_hyperband):
...
+    study = optuna.create_study()
+
+    # Prepare sample from uniform distribution for cheking other distributions.
+    uni_dist = optuna.distributions.UniformDistribution(1.0, 100.0)
+    past_trials = [frozen_trial_factory(i, dist=uni_dist) for i in range(1, 8)]
+    trial = frozen_trial_factory(8)
+    sampler = TPESampler(n_startup_trials=5, seed=0)
+    with patch("optuna.Study.get_trials", return_value=past_trials):
+        uniform_suggestion = sampler.sample_independent(study, trial, "param-a", uni_dist)
+    assert 1.0 <= uniform_suggestion <= 100.0
```
Comment (hvy) [inline tests/samplers_tests/tpe_tests/test_sampler.py:147]: A detail but `high` is expected to be exclusive https://github.com/optuna/optuna/blob/master/optuna/distributions.py#L110 so how about the follwing?
```suggestion
    assert 1.0 <= uniform_suggestion < 100.0
```

File: tests/samplers_tests/tpe_tests/test_sampler.py [hunk 5]
```diff
@@ -20,6 +29,281 @@ def test_hyperopt_parameters(use_hyperband):
...
+        uniform_suggestion = sampler.sample_independent(study, trial, "param-a", uni_dist)
+
+    # Test sample from log-uniform is different from uniform.
+    log_dist = optuna.distributions.LogUniformDistribution(1.0, 100.0)
+    past_trials = [frozen_trial_factory(i, dist=log_dist) for i in range(1, 8)]
+    trial = frozen_trial_factory(8)
+    sampler = TPESampler(n_startup_trials=5, seed=0)
+    with patch("optuna.Study.get_trials", return_value=past_trials):
+        loguniform_suggestion = sampler.sample_independent(study, trial, "param-a", log_dist)
+    assert 1.0 <= loguniform_suggestion <= 100.0
```
Comment (hvy) [inline tests/samplers_tests/tpe_tests/test_sampler.py:168]: Ditto.
```suggestion
    assert 1.0 <= loguniform_suggestion < 100.0
```

**Comment by ytsmiling** [comment]
Current implementation samples from [low, high] [(here)](https://github.com/optuna/optuna/blob/ae0c8b483132e70c2ee30ad10f9156cfd4366bba/optuna/samplers/tpe/sampler.py#L331).
While it occurs in probability zero, I'll create a PR that specially handles the case where sampled value is equal to high.

**Comment by hvy** [comment]
Could you merge the latest master? Seems like you need to update the CircleCI configuration.

**Pull request merged**